common/cache/metricsScopeCache.go

```diff
@@ -26,47 +26,113 @@ import (
 	"bytes"
 	"strconv"
 	"sync"
+	"sync/atomic"
+	"time"
 
 	"github.com/uber/cadence/common/metrics"
 )
 
-type domainMetricsScopeCache struct {
-	sync.RWMutex
-	scopeMap map[string]metrics.Scope
-}
+const flushBufferedMetricsScopeDuration = 10 * time.Second
+
+type (
+	metricsScopeMap map[string]metrics.Scope
+
+	buffer struct {
+		sync.RWMutex
+		bufferMap metricsScopeMap
+	}
+
+	domainMetricsScopeCache struct {
+		buffer   *buffer
+		cache atomic.Value
+		closeCh chan int
+	}
+)
 
 // NewDomainMetricsScopeCache constructs a new domainMetricsScopeCache
-func NewDomainMetricsScopeCache() DomainMetricsScopeCache {
-	return &domainMetricsScopeCache{
-		scopeMap: make(map[string]metrics.Scope),
+func NewDomainMetricsScopeCache(flushDuration ...time.Duration) DomainMetricsScopeCache {
+
+	mc := new(domainMetricsScopeCache)
+	mc.buffer = &buffer{
+		bufferMap: make(metricsScopeMap),
+	}
+	mc.closeCh = make(chan int)
+	mc.cache.Store(make(metricsScopeMap))
+
+	var refreshDuration time.Duration
+	if len(flushDuration) == 0 {
+		refreshDuration = flushBufferedMetricsScopeDuration
+	} else {
+		refreshDuration = flushDuration[0]
+	}
+
+	go mc.flushBufferedMetricsScope(refreshDuration)
+	return mc
+}
+
+func (c *domainMetricsScopeCache) flushBufferedMetricsScope(flushDuration time.Duration) {
+	for {
+		select {
+		case <-time.After(flushDuration):
+			c.buffer.Lock()
+			if len(c.buffer.bufferMap) > 0 {
+				scopeMap := make(metricsScopeMap)
+
+				data := c.cache.Load().(metricsScopeMap)
+				// Copy everything over after atomic load
+				for key, val := range data {
+					scopeMap[key] = val
+				}
+
+				// Copy from buffered array
+				for key, val := range c.buffer.bufferMap {
+					scopeMap[key] = val
+				}
+
+				c.cache.Store(scopeMap)
+				c.buffer.bufferMap = make(metricsScopeMap)
+			}
+			c.buffer.Unlock()
+
+		case <- c.closeCh:
+			break
```

Comment:
I don't think this will work. `break` here will exit the `select` statement, not the for loop. https://golang.org/ref/spec#Break_statements